### DISCUSSION TITLE
Defining a language with AST

### DISCUSSION THREAD

**Discussion by LegNeato** [egraphs-good/egg#95] [Q&A] [answered]
Hello! Cool lib. I'm looking to use egg to transform some source code I have. I have the AST parsed in Rust. How do I take the AST and use it with egg? Will it be easier parse the source with egg instead?

Sorry, I am a bit lost even after looking at the examples.

**Comment by mwillsey** [comment, marked as answer]
Thanks! egg requires pretty much total control over your AST, since it needs to be able to put stuff in the e-graph. So it's pretty common to have an "algebraic" version of your AST that you'll work with _and_ and egg version that's flatter.

So yours might look like
```rust
enum Expr {
  Add(Box<Self>, Box<Self>),
  Num(i32),
}
```

But the egg version will look like 
```rust
define_language! {
    enum SimpleLanguage {
        "+" = Add([Id; 2]),
        Num(i32),
    }
}
```

where basically every instance of recursion in the AST is replaced with an `Id`.

**Reply by LegNeato** [reply]
Ah ok, that's what I suspected but thought there might be a trait or conversion I could implement on my ast

**Reply by mwillsey** [reply]
I would love to have such a trait, but I have tried and failed to design one that does what you describe. I think we need higher kinded types or something.
